Why do we have final size? (#3739)
quicwg/base-drafts: draft-ietf-quic-transport.md

than the offset of the byte with the largest offset sent on the stream, or zero
if no bytes were sent.

At the end of a stream, the sender communicates the final size to the recipient,
ensuring flow control state is synchronized.  For a stream that is reset, the
final size is carried explicitly in a RESET_STREAM frame.  Otherwise, the final
size is the offset plus the length of a STREAM frame marked with a FIN flag, or

> 0 in the case of incoming unidirectional streams.

Reviewer: To me it seems that the text starting from "or, ..." is confusing. That is because a "sender" does not communicate the final size of incoming undirectional stream, contrary to what the first sentence of this paragraph suggests.

I think it might make sense to either simply drop "or... ", or modify the first sentence to state that such communication is made for streams that has the send side.

Reviewer 2: I think that Kazuho is right regarding unidirectional streams (implying that a nothingness has size is silly), but I would go further.

There are three cases we care about: STREAM w/ data and FIN in which case the final size is offset+length.  We include an offset even when there is no data so that the final size is communicated in that frame.  Then there is RESET_STREAM, which carries the final size.  That way, no matter how the stream transitions to closed, the size is known.  So I would say that:

> The final size of a stream is always signaled to recipient.  Final size is signaled as the sum of the Offset and Length fields of a STREAM frame, noting that the length might be implicit.  Alternatively, the Final Size field of a RESET_STREAM frame carries this value.  This ensures that all ways that a stream can be closed result in the number of bytes on the stream are reliably transmitted. This guarantees that both endpoints agree on how much flow control credit was consumed by the stream.

Author: Mostly took Martin's text.

than the offset of the byte with the largest offset sent on the stream, or zero
if no bytes were sent.

The final size of a stream is always signaled to the recipient.  The final size
is the sum of the Offset and Length fields of a STREAM frame with a FIN flag,

> noting that the length might be implicit.  Alternatively, the Final Size field

Reviewer: Maybe "noting that the offset and the length might be implicit"?

This is an editorial nitpick purely from the viewpoint of consistency, but it is a bit surprising to see only one of the two fields called out as might be implicit, even though both of the fields can be omitted.

Author: True.  That will only be the case when a stream has only one STREAM frame, so it would be _unusual_ for the final STREAM frame to have an implicit Offset, but it's certainly not prohibited.

> The final size of a stream is always signaled to the recipient.  The final size
> is the sum of the Offset and Length fields of a STREAM frame with a FIN flag,
> noting that these fields might be implicit.  Alternatively, the Final Size field
> of a RESET_STREAM frame carries this value.  This ensures that all ways that a
> stream can be closed result in the number of bytes on the stream being reliably

Reviewer 3: I don't think this is quite correct, since there's not an implicit fin in the case of a stream where the Fin was never sent and the stream was reset.

```suggestion
The final size of a stream is always signaled to the recipient.  For streams that
are completed, the final size is the sum of the Offset and Length fields of a
STREAM frame with a FIN flag.  For a stream that is reset, the Final Size field
of a RESET_STREAM frame carries the final size.  This ensures that all ways that a
stream can be closed result in the number of bytes on the stream being reliably
```

Reviewer 2: The implicit refers to the Length field, which is implicit when omitted.  Some tweaking might be required for that.

Reviewer 4: No need to refer to the wire encoding of offset and length, right?
My suggestion: "... is the sum of the offset and the length of a STREAM frame with a FIN flag"

Author: Both the Offset and Length fields can be implicit.  As discussed above, it would be unusual (at least with HTTP/3 for both to be implicit, but it's certainly allowed.  I think if the fields specifically get mentioned, the aside that they might be implicit needs to remain.  If we switch to pointing to the values, we could drop that, but it might be confusing to some people.

FIN isn't a field, and isn't contemplated by the aside about implicit fields.

> of a RESET_STREAM frame carries this value.  This ensures that all ways that a
> stream can be closed result in the number of bytes on the stream being reliably
> transmitted.  This guarantees that both endpoints agree on how much flow control

Reviewer 3: ```suggestion
of a RESET_STREAM frame carries this value.  This ensures all closed streams have a known final size, guaranteeing both endpoints agree on how much flow control
```
